wurst/objects/abilities/beastMaster/Pets/PetMovement.wurst

```diff
@@ -0,0 +1,120 @@
+import LinkedList
+
+// Local Imports:
+import LocalObjectIDs
+import Pets
+import Classes
+
+// Array of the mouse positions kept for moving the pet in manual control mode.
+vec2 array MOUSE_POS
+
+// Map of players to their chosen control scheme for pets.
+public let PET_CONTROLS = new HashMap<player, bool>()
+
+// Radius that pet will search for enemy when moved in manual control.
+let PET_ATTACK_RADIUS = 100.
+
+// Pet will randomly choose a position in this radius from target position in automatic mode.
+// This is done to prevent the pet and troll from stacking as well as an aesthetic choice.
+let PET_MOVE_RANGE = 200.
+
+// The orders that the pet is allowed to copy using auto movement.
+let ALLOWED_ORDERS = asList(
+    Orders.smart,
+    Orders.move,
+    Orders.attack,
+    Orders.stop,
+    Orders.holdposition
+)
+
+// Used to continuously track the mouse position of players, used to order pet in manual control mode.
+function updateMousePos()
+    MOUSE_POS[EventData.getTriggerPlayer().getId()] = EventData.getMouseWorldPos()
+
+// Manual control pet movement.
+function onKeyDown(player triggerer)
+    // Exit if the player is not using manual control mode.
+    if PET_CONTROLS.get(triggerer) == true
+        return
+    
+    // Get the triggerer's Pet.
+    let pet = triggerer.getPet() 
+
+    // Find the nearest enemy in range of the target position.
+    forNearestUnit(MOUSE_POS[triggerer.getId()], PET_ATTACK_RADIUS, Filter(-> GetFilterUnit().isEnemyOf(GetTriggerPlayer()))) (unit u) ->
+        // If an enemy is found.
+        if u != null and u.isEnemyOf(triggerer)
+            // Attack the enemy
+            pet.issueTargetOrderById(Orders.attack, u)
+
+        // If no enemy is found.
+        else 
+            // Move to target location.
+            pet.issuePointOrderById(Orders.smart, MOUSE_POS[triggerer.getId()])
+
+// Auto control pet movement.
+function autoMovePet()
+    // Exit if the player is using manual control mode.
+    if PET_CONTROLS.get(EventData.getTriggerPlayer()) == false
+        return
+
+    // Store the moving unit.
+    let u = EventData.getTriggerUnit()
+
+    // Verify the moving unit has a pet and is a derivitive of Beast Master.
+    if u.getOwner().getPet() != null and getTrollBaseClass(u) == UNIT_BEAST_MASTER
+        // Store the pet of moving unit.
+        let pet = u.getOwner().getPet()
+
+        let prevOrder = u.getLastOrder()
+
+        // Compute the type of the previous order.
+        let orderType = prevOrder == null
+            ? OrderType.IMMEDIATE
+            : prevOrder.orderType
+
+        // Exit if the pet is not allowed to copy the order.
+        if not ALLOWED_ORDERS.has(prevOrder.orderId)
+            return
+        
+        // Issue the previous order to the pet.
+        switch orderType
+            case POINT
+                // Moves the pet to a random location in a small circle around target pos.
+                vec2 targetPos = prevOrder.targetPosition
```

Comment:
I've noticed you're using multiple lines for offsetting a `vec2` in several places. Tuples can support direct operators and `vec2` does specifically. You can use `vec2(1, 3) + vec2(5, 6)` for example with no overhead and simplify some of this syntax.